This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [6]:
#Note: needed to install load_dotenv into the terminal

# get api key from your .env file
import os
from dotenv import load_dotenv

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

print(API_KEY)

zzdUa37nkZWyQ_u9xzzh


Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [14]:
# First, import the relevant modules

import requests

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [73]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned

url = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X/data.json?&rows=1&api_key=zzdUa37nkZWyQ_u9xzzh'
r = requests.get(url)
json_data = r.json()



# curl "https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2023-02-02&end_date=2023-02-02&api_key=zzdUa37nkZWyQ_u9xzzh"

In [74]:
# print(json_data)

for k in json_data.keys():
    print(k + ': ', json_data[k])

dataset_data:  {'limit': 1, 'transform': None, 'column_index': None, 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'start_date': '2000-06-07', 'end_date': '2020-12-01', 'frequency': 'daily', 'data': [['2020-12-01', 112.2, 112.2, 111.5, 112.0, None, 51.0, 5703.0, None, None, None]], 'collapse': None, 'order': None}


In [4]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure

{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-12-01T14:48:09.907Z',
  'newest_available_date': '2020-12-01',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2021-01-03',
  'end_date': '2020-12-01',
  'data': [],
  'collapse': None,
  'order': None,
  'database_id': 6129}}

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [142]:
#Task 1

url_task = 'https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-12-31&api_key=zzdUa37nkZWyQ_u9xzzh'
r_task = requests.get(url_task)

In [143]:
#Task 2
json_data_task = r_task.json()

# for k in json_data_task.keys():
#     print(k + ': ', json_data_task[k])

In [146]:
#Task 3. Calculate what the highest and lowest opening prices were for the stock in this period.
column_names = json_data_task['dataset_data']['column_names']
data = json_data_task['dataset_data']['data']
print(column_names)
# print(data)


['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover']


In [193]:
#Task 3 Cont.

# Method 1
highest_open = data[0][1]
lowest_open = data[0][1]

for day in data :
    if day[1] != None :
        if day[1] > highest_open :
            highest_open = day[1]
        if day[1] < lowest_open :
            lowest_open = day[1]

print(highest_open)
print(lowest_open)

# Method 2
# dates = []
# opens = []

# for day in data :
#     dates.append(day[0])
#     opens.append(day[1])

# print(opens)

53.11
34.0


In 2017, the highest open was 53.11, and the lowest open was 34.0

In [195]:
#Task 4. What was the largest change in any one day (based on High and Low price)?
largest_change = data[0][2] - data[0][3]

for day in data :
    if day[2] != None and day[3] != None :
        change = day[2] - day[3]
        if change > largest_change :
            largest_change = change

print(largest_change)

2.8100000000000023


In 2017, the largest swing within one day was 2.81.

In [206]:
#Task 5: What was the largest change between any two days (based on Closing Price)?

closing_prices = []
for day in data :
    closing_prices.append(day[4])

# print(closing_prices)

largest_close_change = abs(closing_prices[1] - closing_prices[0])

diffs_closing_price = []

for n in range(len(closing_prices)-1) :
    diff = abs(closing_prices[n+1] - closing_prices[n])
    diffs_closing_price.append(diff)
    if diff > largest_close_change :
        largest_close_change = diff

print(largest_close_change)
# print(diffs_closing_price)




2.559999999999995


In 2017, the largest change between any two days (based on Closing Price) was 2.56.

In [208]:
# Task 6. What was the average daily trading volume during this year?
trading_volume = []
for day in data :
    trading_volume.append(day[6])

avg = sum(trading_volume) / len(trading_volume)
print(avg)

[34640.0, 40660.0, 57452.0, 71165.0, 120649.0, 50587.0, 137313.0, 129733.0, 204080.0, 132981.0]
89124.33725490196


In 2017, the average daily trading volume was ~80,124 shares.

In [ ]:
# 7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

#Not doing this for now